silver/models/documents/base.py

```diff
@@ -493,3 +527,99 @@ def total_before_tax_in_transaction_currency(self):
     def tax_value_in_transaction_currency(self):
         return sum([entry.tax_value_in_transaction_currency
                     for entry in self.entries])
+
+    @property
+    def amount_paid_in_transaction_currency(self):
+        Transaction = get_model('silver.Transaction')
+
+        return sum([transaction.amount
+                    for transaction in self.transactions.filter(state=Transaction.States.Settled)])
+
+    @property
+    def amount_pending_in_transaction_currency(self):
+        Transaction = get_model('silver.Transaction')
+
+        return sum([transaction.amount
+                    for transaction in self.transactions.filter(state=Transaction.States.Pending)])
+
+    @property
+    def amount_to_be_charged_in_transaction_currency(self):
+        Transaction = get_model('silver.Transaction')
+
+        return self.total_in_transaction_currency - sum([
+            transaction.amount
+            for transaction in self.transactions.filter(state__in=[
+                Transaction.States.Initial,
+                Transaction.States.Pending,
+                Transaction.States.Settled
+            ])
+        ])
+
+
+def create_transaction_for_document(document):
+    # get a usable, recurring payment_method for the customer
+    PaymentMethod = get_model('silver.PaymentMethod')
+    Transaction = get_model('silver.Transaction')
+
+    payment_methods = PaymentMethod.objects.filter(
+        canceled=False,
+        verified=True,
+        customer=document.customer
+    )
+    for payment_method in payment_methods:
+        try:
+            return Transaction.objects.create(document=document,
+                                              payment_method=payment_method)
+        except ValidationError:
+            continue
+
+
+@receiver(post_transition)
+def post_transition_callback(sender, instance, name, source, target, **kwargs):
+    if not isinstance(instance, BillingDocumentBase):
+        return
+
+    document = instance
+    setattr(document, '.recently_transitioned', target)
+
+    document.save()
+
+
+@receiver(post_save)
+def post_document_save(sender, instance, created=False, **kwargs):
+    if not isinstance(instance, BillingDocumentBase):
+        return
+
+    document = instance
+
+    if hasattr(document, '.recently_transitioned'):
+        delattr(document, '.recently_transitioned')
+
+        # Transition related document too, if needed
+        if document.related_document and document.state != document.related_document.state:
+            state_transition_map = {
+                BillingDocumentBase.STATES.ISSUED: 'issue',
+                BillingDocumentBase.STATES.CANCELED: 'cancel',
+                BillingDocumentBase.STATES.PAID: 'pay'
+            }
+            transition_name = state_transition_map[document.state]
+
+            bound_transition_method = getattr(document.related_document, transition_name)
+            bound_transition_method()
```

Comment:
1. Yes, it will also save the related document, because the `save` method gets automatically called after a successful transition.
2. I'd rather have this happen synchronously. Having a draft proforma and an issued invoice, even for a short period of time, doesn't sound good.